src/compute/src/logging/timely.rs

```diff
@@ -526,3 +325,406 @@ fn create_address_row(id: usize, worker: WorkerIdentifier, address: &[usize]) ->
+
+            let datum = AddressDatum {
+                id: channel.id,
+                address: channel.scope_addr,
+            };
+            self.output.addresses.give(self.cap, (datum, ts, -1));
+
+            // Retract messages logged for this channel.
+            if let Some(sent) = self.state.messages_sent.remove(&channel.id) {
+                for (target_worker, count) in sent.iter().enumerate() {
+                    let datum = MessageDatum {
+                        channel: channel.id,
+                        worker: target_worker,
+                    };
+                    self.output
+                        .messages_sent
+                        .give(self.cap, (datum, ts, -count));
+                }
+            }
+            if let Some(received) = self.state.messages_received.remove(&channel.id) {
+                for (source_worker, count) in received.iter().enumerate() {
+                    let datum = MessageDatum {
+                        channel: channel.id,
+                        worker: source_worker,
+                    };
+                    self.output
+                        .messages_received
+                        .give(self.cap, (datum, ts, -count));
+                }
+            }
+        }
+    }
+
+    fn handle_park(&mut self, event: ParkEvent) {
+        let time_ns = self.time.as_nanos();
+        match event {
+            ParkEvent::Park(requested) => {
+                let park = Park { time_ns, requested };
+                let existing = self.state.last_park.replace(park);
+                if existing.is_some() {
+                    error!("park without a succeeding unpark");
+                }
+            }
+            ParkEvent::Unpark => {
+                if let Some(park) = self.state.last_park.take() {
+                    let duration_ns = time_ns - park.time_ns;
+                    let duration_pow = duration_ns.next_power_of_two();
+                    let requested_pow = park.requested.map(|r| r.as_nanos().next_power_of_two());
+
+                    let ts = self.ts();
+                    let datum = ParkDatum {
+                        duration_pow,
+                        requested_pow,
+                    };
+                    self.output.parks.give(self.cap, (datum, ts, 1));
+                } else {
+                    error!("unpark without a preceeding park");
+                }
+            }
+        }
+    }
+
+    fn handle_messages(&mut self, event: MessagesEvent) {
+        let ts = self.ts();
+        let count = Diff::try_from(event.length).expect("must fit");
+
+        if event.is_send {
+            let datum = MessageDatum {
+                channel: event.channel,
+                worker: event.target,
+            };
+            self.output.messages_sent.give(self.cap, (datum, ts, count));
+
+            let sent_counts = self.state.messages_sent.entry(event.channel).or_default();
+            grow_vec(sent_counts, event.target);
```

Comment:
The previous version of the code was pre-allocating this to `peers`. Now, we will grow the vector as we see higher indexes over time. Not sure if we are worried about the performance implications here, as this is done for every channel. It _seems_ to be fine, as we should grow the vector pretty quickly upon seeing different targets and stabilize.

Perhaps more importantly, this version will not panic if we are given a `target` that is inconsistent with `peers`. I am fine with increasing availability and avoid crashing the replica, but it would be a good idea to get this checked by someone with more knowledge of Timely's logging semantics. We might not want to keep executing in a bad state.

Reply:
I did run a simple performance test with this branch and it seemed to be even faster than main for some reason, so I'm not worried about the performance aspects of this change. The panicking aspect might be a cause for concern indeed. I'm happy to change it back. That'd just mean threading the `peers` count to the `DemuxHandler` too.

Reply:
Perhaps @antiguru can give a final verdict here too?

Reply:
Let's just go with the `peers` approach! That way we don't have to feel bad about leaving performance on the floor and we can be sure to get to know about messed-up timely states.